### PR TITLE
Localization

### LINKED ISSUE #17: Make "Edit with Notepad++" translatable
NICE-TO-HAVE: It'll be great to make "Edit with Notepad++" translatable.

### PR DESCRIPTION
**Pull request by GurliGebis**
This changes the "Edit with Notepad++" string to be loaded from the resources instead of being hardcoded.

I have translated it to danish (since that is what I speak).
The french translation is taken from [here](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13201#issuecomment-1499436949)

This fixes #17

### CODE REVIEW

**Comment by GurliGebis** [comment]
@donho this one is ready for review - it falls back to en-US as the message, if the current UI language isn't added to the DLL.

**Comment by FadeMind** [comment]
@GurliGebis please add Polish language to this PR. 

IBM Code for Polish language is `PLK`
based on https://www.ibm.com/docs/en/cics-ts/5.3?topic=development-national-language-codes

```
/////////////////////////////////////////////////////////////////////////////
// Polski (Polska) resources

#if !defined(AFX_RESOURCE_DLL) || defined(AFX_TARG_PLK)
LANGUAGE LANG_POLISH, SUBLANG_DEFAULT

/////////////////////////////////////////////////////////////////////////////
//
// String Table
//

STRINGTABLE
BEGIN
    IDS_EDIT_WITH_NOTEPADPLUSPLUS "Edytuj w Notepad++"
END

#endif    // Polski (Polska) resources
/////////////////////////////////////////////////////////////////////////////
```

**Comment by GurliGebis** [comment]
@FadeMind I have added all the translations from the old NppShell_06 dll, so Polish is there as well 🙂
@donho there are all there now, so it is ready.

File: BaseNppExplorerCommandHandler.cpp [hunk 1]
```diff
@@ -21,7 +24,17 @@ const wstring BaseNppExplorerCommandHandler::GetNppExecutableFullPath()
...
 {
-    return L"Edit with Notepad++";
+    // A buffer size of 1024 should be enough to hold the for all languages.
+    constexpr int bufferSize = 1024;
+
+    // Buffer to store the string resource into.
+    WCHAR buffer[bufferSize];
+
+    // Load the string from the resource matching the current language.
+    LoadStringW(thisModule, IDS_EDIT_WITH_NOTEPADPLUSPLUS, buffer, bufferSize);
```
Comment (donho) [inline BaseNppExplorerCommandHandler.cpp:34]: Ho to determinate which `IDS_EDIT_WITH_NOTEPADPLUSPLUS` we should load?
It seems to me on the compile time, that means that we have to compile several NppShell.dll correspondent to different languages, then we deploy the right one during the installation? Or am I missing somethings?

**Review by donho — CHANGES_REQUESTED** [review]
Please provide the answer for the question.

File: BaseNppExplorerCommandHandler.cpp [hunk 2]
```diff
@@ -1,11 +1,14 @@
 #include "pch.h"
+#include "resource.h"
 #include "BaseNppExplorerCommandHandler.h"
 
 #include "PathHelper.h"
 
 using namespace NppShell::CommandHandlers;
 using namespace NppShell::Helpers;
 
+extern HMODULE thisModule;
```
[the remark this answers is not in the document]
Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp:10]: I will rename it once I get home.

File: BaseNppExplorerCommandHandler.cpp (hunk 1 above)
Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp:34]: LoadStringW takes from the string table with the matching language (defaults to en-US if none is found).
So just this one dll, and it just works (we could move it out to separate NppShell.dll.mui files, but that complicated it)
It is pretty nice that Windows figures out which table to read from automatically 😃

Comment (donho) [reply BaseNppExplorerCommandHandler.cpp:34]: > LoadStringW takes from the string table with the matching language (defaults to en-US if none is found).

How does LoadStringW match language? With the OS language?
 
If yes, how a user under Japanese Windows can disable the translation and restore it to "Edit with Notepad++"?

Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp:34]: It matches the UI language of Windows (so it will match the rest of the list items the user is seeing).
Why would a user with a Japanese version of Windows want to force the English text version, when the rest of the context menu is in Japanese? (Cut, Copy, Paste etc.)
I mean, if they don't understand Japanese, they wouldn't be using the Japanese version of the Windows UI (I would assume), and if they install the Japanese version of Windows, and then use the English user interface pack, then it will follow that.

So in short, it will be the same language as the rest of the context menu.
Which is what I think is the best solution, since it gives the best integration (matching the rest of the system is kind of "just works") 🙂

File: BaseNppExplorerCommandHandler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp]: Renamed in ed3f1ba51a9dcf4929797fbdca957e34861f2659

File: BaseNppExplorerCommandHandler.cpp (hunk 1 above)
Comment (donho) [reply BaseNppExplorerCommandHandler.cpp]: > So in short, it will be the same language as the rest of the context menu.

Not necessarily. Even on my French Windows, I prefer to use software in English.
And a lot of my colleagues do the same. 
In a enterprise environment, the default language of installed Windows matches usually to the local language, but people could want to have their software in English.

Using an external localization file (or even storing the translation in the registry) allows users do whatever they want.

Comment (donho) [reply BaseNppExplorerCommandHandler.cpp]: However, I prefer the simplicity of current code of this PR. 
It could be kept if we can allow user under other localization windows - for example if a empty xml file "useDefaultEnlishTitle.xml"  is found beside of NppShell.dll, the function returns "Edit with Notepad++". What do you think ?

Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp]: Sure, I can easily add logic to look for that file and force english.
I'll look at it this weekend and update this PR

**Comment by GurliGebis** [comment]
@donho done.
If a file named "overrideMenuEntryLanguage.txt" exists in the folder, the following will happen:
- If the file is empty, it will use "en-US" as language.
- If the file contains a language (in the screenshot, it is da-DK), it will use that.

screenshot

Now it should be ready 🙂

File: BaseNppExplorerCommandHandler.cpp (hunk 1 above)
Comment (GurliGebis) [reply BaseNppExplorerCommandHandler.cpp]: @donho see below: https://github.com/notepad-plus-plus/nppShell/pull/25#issuecomment-1500951534

**Comment by GurliGebis** [comment]
That way we can override it to whatever language the user wants
